Add QueueConsumer component for Cloudflare Queues

bachiitter opened the pull request:
This PR adds `QueueConsumer` component for Cloudflare Queues allowing user to separate queue message processing into dedicated workers.

## Example Usage
```ts
const queue = new sst.cloudflare.Queue("MyQueue");

new sst.cloudflare.QueueConsumer("MyConsumer", {
  queueId: queue.id,
  processor: "consumer.ts"
});

bachiitter committed:
feat: add QueueConsumer component for Cloudflare Queues

bachiitter committed:
Merge branch 'anomalyco:dev' into feature/queue-consumer

vimtor reviewed (commented):
thanks for your contribution @bachiitter 

i'm thinking that we should probably follow the same API as the AWS `queue.subscribe()` instead of exposing the queue consumer directly

also, it would be amazing if you could add an example in the `examples/` directory

what do you think?

bachiitter commented:
> thanks for your contribution @bachiitter
> 
> i'm thinking that we should probably follow the same API as the AWS `queue.subscribe()` instead of exposing the queue consumer directly
> 
> also, it would be amazing if you could add an example in the `examples/` directory
> 
> what do you think?

will create an example, queue.subscribe() seems better though i don't have much experience so what ever fits better

vimtor commented:
yep, i prefer `queue.subscribe` since it's more consistent with what we already have

bachiitter committed:
feat(cloudflare): add subscribe method to Queue component

- Move QueueConsumer functionality into Queue.subscribe()
- Remove separate QueueConsumer component
- Add cloudflare-queue example
- Add QueueConsumer to component exemption list

Amp-Thread-ID: https://ampcode.com/threads/T-019c15f2-4742-72d0-93de-a02a1d0f37d2
Co-authored-by: Amp <amp@ampcode.com>

bachiitter commented:
@vimtor updated to queue.subscribe and also added an example

vimtor commented on examples/cloudflare-queue/producer.ts:
```diff
@@ -0,0 +1,8 @@
+
+export default {
+  async fetch(request, env) {
+    await env.MyQueue.send({ hello: "world" });
```
shouldn't we use this instead?

```suggestion
    await Resource.MyQueue.send({ hello: "world" });
```

vimtor commented on platform/src/components/cloudflare/queue.ts:
```diff
@@ -49,6 +129,96 @@ export class Queue extends Component implements Link.Linkable {
+    opts?: ComponentResourceOptions,
+  ) {
+    if (this.isSubscribed) {
+      throw new Error(
```
we should throw a `VisibleError` type so it gets printed

```suggestion
      throw new VisibleError(
```

vimtor commented on platform/src/components/cloudflare/queue.ts:
```diff
@@ -49,6 +129,96 @@ export class Queue extends Component implements Link.Linkable {
+        ),
+      );
+
+      return { worker, consumer };
```
not sure about this return type

to be consistent with the aws queue we should create a `QueueWorkerSubscriber` component and return that instead

vimtor reviewed (changes requested):
looking great @bachiitter!

i've left a couple of comments with some minor changes

we should be ready to merge this after fixing them

bachiitter commented:
> looking great @bachiitter!
> 
> i've left a couple of comments with some minor changes
> 
> we should be ready to merge this after fixing them

just copy example from another test project will update it and will make other fixes

bachiitter committed:
fix example to use Resource instead of env

bachiitter committed:
switch to visible error

bachiitter committed:
extract QueueWorkerSubscriber component

vimtor reviewed (commented):
<details>

<summary>your code works like a charm @bachiitter</summary>

<img width="1526" height="596" alt="Image" src="https://github.com/user-attachments/assets/7b946eba-e285-46e2-8055-bb2571c84c58" />

<img width="2974" height="936" alt="Image" src="https://github.com/user-attachments/assets/b61fad69-956a-49a5-a8fb-63a70f9aef79" />

</details>

however it gives the following error when removing the deployment:

```
MyQueueSubscriber sst:cloudflare:QueueWorkerSubscriber → MyQueueSubscriberConsumer cloudflare:index:QueueConsumer
failed to make http request: missing required consumer_id parameter
```

this comes from the cloudflare provider https://github.com/cloudflare/terraform-provider-cloudflare/issues/6405

so we'll have to wait for it to get fixed before merging

bachiitter commented:
> your code works like a charm @bachiitter
> <img alt="Image" width="1526" height="596" src="https://private-user-images.githubusercontent.com/36263538/545775126-7b946eba-e285-46e2-8055-bb2571c84c58.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzAzMTcxOTEsIm5iZiI6MTc3MDMxNjg5MSwicGF0aCI6Ii8zNjI2MzUzOC81NDU3NzUxMjYtN2I5NDZlYmEtZTI4NS00NmUyLTgwNTUtYmIyNTcxYzg0YzU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMjA1VDE4NDEzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNkNDVkMWUyOTgxYWQ3NjIxYTFkNGE1ODM0MDc2ODk5ZWUzZGMwODYwNmQwNjEzZTE4ODM1YTQ4NWI3MTNmYzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.T15Hxw57SIO7HKNyA1FsUXVCndPyxlHcsOOA6PbYONA"> <img alt="Image" width="2000" height="936" src="https://private-user-images.githubusercontent.com/36263538/545775139-b61fad69-956a-49a5-a8fb-63a70f9aef79.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzAzMTcxOTEsIm5iZiI6MTc3MDMxNjg5MSwicGF0aCI6Ii8zNjI2MzUzOC81NDU3NzUxMzktYjYxZmFkNjktOTU2YS00OWE1LWE4ZmItNjNhNzBmOWFlZjc5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMjA1VDE4NDEzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA2ZmVkYTNiMDYzZjA3YmY0NTU4NjBlYTc3NTNlOGUwYmQ5NmJhMTE3YTdjNTQyMjY0OTBhN2I2ZTk4MDJlZDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.YERhaN-1ndU9Fv3FNFFSdBOko3n43EI5QGLHy5bu-jE">
> however it gives the following error when removing the deployment:
> 
> ```
> MyQueueSubscriber sst:cloudflare:QueueWorkerSubscriber → MyQueueSubscriberConsumer cloudflare:index:QueueConsumer
> failed to make http request: missing required consumer_id parameter
> ```
> 
> this comes from the cloudflare provider [cloudflare/terraform-provider-cloudflare#6405](https://github.com/cloudflare/terraform-provider-cloudflare/issues/6405)
> 
> so we'll have to wait for it to get fixed before merging

another thing added to list of thing cloudflare needs fixing

bachiitter committed:
Merge branch 'anomalyco:dev' into feature/queue-consumer

bachiitter committed:
Merge branch 'anomalyco:dev' into feature/queue-consumer

vimtor commented on platform/src/components/cloudflare/queue.ts:
```diff
@@ -18,17 +21,95 @@ export interface QueueArgs {
+  /**
+   * The dead letter queue to send messages that fail processing.
+   */
+  deadLetterQueue?: Input<string>;
```
shouldn't the dlq be an arg to the queue itself?

bachiitter replied:
it's on consumer
https://www.pulumi.com/registry/packages/cloudflare/api-docs/queueconsumer/

vimtor replied:
i see

given we can only have one consumer, i'll probably still put it in the queue just to match the existing aws queue component

what do you think? open to suggestions here

vimtor replied:
@bachiitter thoughts?

bachiitter replied:
sure we can do that only problem is we can't do that for the queue-worker-subscriber

vimtor commented on platform/src/components/cloudflare/queue.ts:
```diff
@@ -18,17 +21,95 @@ export interface QueueArgs {
+  /**
+   * Consumer settings like batch size, retries, and visibility timeout.
+   */
+  settings?: cloudflare.QueueConsumerArgs["settings"];
```
could we directly expose the settings ourselves? similar to how we do it in the aws queue? with nice short names and defaults. ideally we will mirror the same field structure

bachiitter replied:
sure we can do that, will look at aws queue

vimtor commented on examples/cloudflare-queue/bun.lock:
could you remove this file?

(sorry this should be ignored by default)

bachiitter replied:
> could you remove this file?
> 
> (sorry this should be ignored by default)

sure sorry my bad

vimtor reviewed (changes requested):
@bachiitter good news! with the latest cloudflare provider the removing issue no longer happens

i left a couple of minor comments, once resolved we should be good to merge

vimtor commented:
@bachiitter forgot to mention that we should surface it in the docs in this pr as well

bachiitter committed:
remove bun.lock

bachiitter committed:
Align Cloudflare queue subscriber args with Queue.subscribe API

bachiitter committed:
remove git ignore

bachiitter committed:
Merge branch 'dev' into feature/queue-consumer

vimtor commented on .gitignore:
```diff
@@ -3,6 +3,7 @@ dist
 node_modules
 Pulumi.*.yaml
 Pulumi.yaml
+bun.lock
```
sorry we shouldn't ignore this i misspoke 

we should only ignore it for the example 

don't worry i will fix it in a separate pr

bachiitter replied:
got it

vimtor commented on examples/cloudflare-queue/package.json:
```diff
@@ -0,0 +1,10 @@
+  "name": "cloudflare-queue",
+  "version": "0.0.0",
+  "dependencies": {
+    "sst": "^3.1.49"
```
can we set it to ^4 like the other examples?

vimtor commented on platform/src/components/cloudflare/queue-worker-subscriber.ts:
```diff
@@ -0,0 +1,196 @@
+    /**
+     * The name of the dead letter queue.
+     */
+    queue?: Input<string>;
```
when dlq is provided the queue name should be required

bachiitter replied:
got it

vimtor commented on platform/src/components/cloudflare/queue.ts:
```diff
@@ -18,17 +22,172 @@ export interface QueueArgs {
+    /**
+     * The name of the dead letter queue.
+     */
+    queue?: Input<string>;
```
same thing here

bachiitter replied:
will fix it

vimtor commented on www/astro.config.mjs:
```diff
@@ -217,6 +217,8 @@ const sidebar = [
           "docs/component/cloudflare/d1",
           "docs/component/cloudflare/kv",
+          "docs/component/cloudflare/queue",
+          "docs/component/cloudflare/static-site",
```
let's remove this one for the moment

i've created a separate issue of it

bachiitter replied:
got it

bachiitter committed:
update sst version

bachiitter committed:
remove static-site from sidebar

bachiitter committed:
remove bun lock from gitignore

bachiitter committed:
Merge branch 'anomalyco:dev' into feature/queue-consumer

bachiitter committed:
require dlq.queue when dlq is configured

vimtor committed:
refactor: move dlq, visibilityTimeout, maxConcurrency to Queue args

- remove dlq validation checks
- rename toMilliSeconds to toMilliseconds
- align with AWS Queue pattern

vimtor committed:
chore: revert generate.ts to origin/dev version

vimtor committed:
docs: add Cloudflare Queue to docs generation

vimtor committed:
fix: remove visibilityTimeout for worker-type queue consumers

visibilityTimeoutMs is only valid for pull-based (http_pull) consumers,
not for worker-type consumers

vimtor committed:
fix: align defaults with Cloudflare Queue docs

- max_batch_timeout (window): 5s (was 20s)
- max_batch_size limit: 100 (was 10000)
- max_batch_timeout limit: 60s (was 300s)

vimtor committed:
fix: pass provider to QueueWorkerSubscriber

Ensures subscriber uses same provider as parent queue

vimtor committed:
fix: add QueueWorkerSubscriber to docs generation
